### PR TITLE
#696 - Support Multiple Quantity

### PR DESCRIPTION
**Pull request by hani-ibrahim**
fixing #696 

applying new design, still need to fix tests and fix any corner cases that are not handled, I can find only one now which is the screen shot of the placed order is not complete. 

But it is in a good state that you can check @bonkey

### CODE REVIEW

File: AtlasUI/AtlasUI/CheckoutSummaryArticleSelectCollectionViewCell.swift [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        let isSelected = type.idx(selectedArticle: selectedArticle) == idx
+
+        switch type {
+        case .size:
+            let unit = selectedArticle.article.availableUnits[idx]
+            valueLabel.text = unit.size
+            priceLabel.attributedText = priceAttributedString(price: unit.price, originalPrice: unit.originalPrice, isSelected: isSelected)
+        case .quantity:
+            let currency = selectedArticle.price.currency
+            let totalPrice = Money(amount: selectedArticle.price.amount * Decimal(idx + 1), currency: currency)
```
Comment (bonkey) [inline AtlasUI/AtlasUI/CheckoutSummaryArticleSelectCollectionViewCell.swift]: 1. `idx+1` should be a named variable.
1. Be more declarative:
```
func *(lhs: Decimal, rhs: Int) -> Decimal {
    return lhs * Decimal(rhs)
}
```

File: AtlasUI/AtlasUI/CheckoutSummaryArticleSelectCollectionViewCell.swift [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        let attributedString = NSMutableAttributedString(attributedString: priceAttributedString(price: price, isSelected: isSelected))
+        attributedString.append(separatorAttributedString(isSelected: isSelected))
+        attributedString.append(originalPriceAttributedString(originalPrice: originalPrice, isSelected: isSelected))
+        return attributedString
+    }
+
+    private func priceAttributedString(price: Money, isSelected: Bool) -> NSAttributedString {
+        let text = Localizer.format(price: price) ?? ""
+        return NSAttributedString(string: text, attributes: [NSFontAttributeName: UIFont.systemFont(ofSize: 10, weight: UIFontWeightLight),
+                                                             NSForegroundColorAttributeName: isSelected ? .white : UIColor.black])
```
Comment (bonkey) [inline AtlasUI/AtlasUI/CheckoutSummaryArticleSelectCollectionViewCell.swift]: White? To hide it? If so, why not return nil?

I would love to see all those colors in a single place. I guess it's not too late, yet ;-)

Comment (bonkey) [inline AtlasUI/AtlasUI/CheckoutSummaryArticleSelectCollectionViewCell.swift]: And BTW. NSAttributedString. Maybe a little overhead, but not that big one:

```swift
extension String {

    var attributed: NSAttributedString {
        return NSAttributedString(string: self)
    }

}


extension NSAttributedString {

    static func attributes(foregroundColor color: UIColor) -> [String: Any] {
        return [NSForegroundColorAttributeName: color]
    }

    static func attributes(font: UIFont) -> [String: Any] {
        return [NSFontAttributeName: font]
    }

    static func attributes(strikeColor color: UIColor) -> [String: Any] {
        return [
            NSStrikethroughStyleAttributeName: 1,
            NSStrikethroughColorAttributeName: color
        ]
    }

    func foreground(color: UIColor) -> NSAttributedString {
        return self.add(attributes: NSAttributedString.attributes(foregroundColor: color))
    }

    func font(_ font: UIFont) -> NSAttributedString {
        return self.add(attributes: NSAttributedString.attributes(font: font))
    }

    func strike(color: UIColor) -> NSAttributedString {
        return self.add(attributes: NSAttributedString.attributes(strikeColor: color))
    }

    func add(attributes: [String: Any]) -> NSAttributedString {
        let mutable = NSMutableAttributedString(attributedString: self)
        let range = NSRange(location: 0, length: self.length)
        mutable.addAttributes(attributes, range: range)
        return NSAttributedString(attributedString: mutable)
    }

}



typealias Money = Decimal
let price: Money = 10
let originalPrice: Money = 15
let selected = true

let locPrice = "\(price)"
let locOriginalPrice = "\(originalPrice)"

let priceColor: UIColor = selected ? .red : .blue
let sepColor: UIColor = selected ? .gray : .green
let origColor: UIColor = selected ? .magenta : .cyan
let strikeColor: UIColor = .red

let mainFont = UIFont.systemFont(ofSize: 10, weight: UIFontWeightLight)
let secondFont = UIFont.systemFont(ofSize: 10, weight: UIFontWeightLight)

let priceFormatted = locPrice.attributed
    .font(mainFont)
    .foreground(color: priceColor)
let separatorFormatted = " ".attributed
    .font(secondFont)
    .foreground(color: origColor)
let originalPriceFormatted = locPrice.attributed
    .font(mainFont)
    .foreground(color: origColor)
    .strike(color: strikeColor)
```

File: AtlasUI/AtlasUI/CheckoutSummaryLayout.swift [hunk 3]
```diff
@@ -75,10 +75,25 @@ struct OrderPlacedLayout: CheckoutSummaryLayout {
...
+
+struct GuestCheckoutLayout: CheckoutSummaryLayout {
+
+    let navigationBarTitleLocalizedKey: String = "summaryView.title.summary"
+    let showCancelButton: Bool = true
+    let showFooterLabel: Bool = true
+    let showDetailArrow: Bool = true
+    let showGuestStackView: Bool = true
+    let showOrderStackView: Bool = false
+    let allowArticleRefine: Bool = true
```
Comment (bonkey) [inline AtlasUI/AtlasUI/CheckoutSummaryLayout.swift]: All the `: Bool`s and `: String` are redundant here.

File: AtlasUI/AtlasUI/GetArticleDetailsViewController.swift [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        getArticle()
+    }
+
+    private func getArticle() {
```
Comment (bonkey) [inline AtlasUI/AtlasUI/GetArticleDetailsViewController.swift]: `get` always makes me trembling ;-)

File: AtlasUI/AtlasUI/CheckoutSummaryArticleSelectCollectionViewCell.swift (hunk 2 above)
Comment (hani-ibrahim) [inline AtlasUI/AtlasUI/CheckoutSummaryArticleSelectCollectionViewCell.swift]: the white is because the selected state background is orange. so white on orange is good. 
simulator screen shot 2 feb 2017 3 47 38 pm

Comment (bonkey) [inline AtlasUI/AtlasUI/CheckoutSummaryArticleSelectCollectionViewCell.swift]: Ah, actually thought about it could be this reason, too, but when I didn't see anything about it in the methods, then I forgot. This means, that this "color theming" should be better grouped somewhere.

**Comment by bonkey** [comment]
All right, seems like rest is pretty fine, but it's also pretty big, so I could have missed something...

**Comment by bonkey** [comment]
👍

File: AtlasUI/AtlasUI/CheckoutSummaryArticleSelectCollectionViewCell.swift (hunk 2 above)
Comment (hani-ibrahim) [inline AtlasUI/AtlasUI/CheckoutSummaryArticleSelectCollectionViewCell.swift]: > This means, that this "color theming" should be better grouped somewhere.

 like an enum for example :japanese_ogre:

File: AtlasUI/AtlasUI/CheckoutSummaryLayout.swift [hunk 5]
```diff
@@ -13,40 +13,40 @@ protocol CheckoutSummaryLayout {
     var showDetailArrow: Bool { get }
     var showGuestStackView: Bool { get }
     var showOrderStackView: Bool { get }
+    var allowArticleRefine: Bool { get }
```
Comment (hani-ibrahim) [inline AtlasUI/AtlasUI/CheckoutSummaryLayout.swift]: `enable`, `allow`, `show` all these are bi-directional

Comment (hani-ibrahim) [inline AtlasUI/AtlasUI/CheckoutSummaryLayout.swift]: `allowsRefining` is very generic however it's for article only

File: AtlasUI/AtlasUI/CheckoutSummaryArticleSelectCollectionViewCell.swift (hunk 2 above)
Comment (bonkey) [inline AtlasUI/AtlasUI/CheckoutSummaryArticleSelectCollectionViewCell.swift]: Why not actually

File: AtlasUI/AtlasUI/CheckoutSummaryLayout.swift (hunk 5 above)
Comment (bonkey) [inline AtlasUI/AtlasUI/CheckoutSummaryLayout.swift]: That's why I suggested it with `s`. It's like a feature it supports, not a switchable parameter.

**Comment by bonkey** [comment]
👍

**Comment by hani-ibrahim** [comment]
👍

File: AtlasUI/AtlasUI/CheckoutSummaryLayout.swift (hunk 5 above)
Comment (hani-ibrahim) [inline AtlasUI/AtlasUI/CheckoutSummaryLayout.swift]: oh yeah adding `s` makes more sense. it also could for other properties as `shows` instead of `show`
will do it in another PR
